sawtooth-sdk-transaction-processor/src/test/java/sawtooth/sdk/processor/TransactionProcessorTest.java

```diff
@@ -0,0 +1,161 @@
+    assertTrue("TransactionProcessor should find the same handler we used before, since we added it again!",
+        handler2.equals(testHandler));
+
+    header = TransactionHeader.newBuilder().setFamilyName(testFamily + "someother")
+        .setFamilyVersion(testVersion + "someother").build();
+    req = TpProcessRequest.newBuilder().setHeader(header).build();
+    Message msgNewFamily = Message.newBuilder().setMessageType(MessageType.TP_PROCESS_REQUEST)
+        .setContent(req.toByteString()).build();
+    testHandler = tp.findHandler(msgNewFamily);
+    assertNull("TransactionProcessor should not find a handler for this message with a new family!", testHandler);
+
+    header = TransactionHeader.newBuilder().setFamilyName(testFamily).setFamilyVersion(testVersion + "someother")
+        .build();
+    req = TpProcessRequest.newBuilder().setHeader(header).build();
+    Message msgNewVersion = Message.newBuilder().setMessageType(MessageType.TP_PROCESS_REQUEST)
+        .setContent(req.toByteString()).build();
+    testHandler = tp.findHandler(msgNewVersion);
+    assertNull("TransactionProcessor should not find a handler for this message with an old family but new version!",
+        testHandler);
+  }
+
+  @Test
+  public void testRun() {
+    Stream stream = mock(Stream.class);
+    TransactionProcessor tp = new TransactionProcessor(stream);
+
+    TransactionHeader header = TransactionHeader.newBuilder().setFamilyName(testFamily).setFamilyVersion(testVersion)
+        .build();
+    TpProcessRequest req = TpProcessRequest.newBuilder().setHeader(header).build();
+    Message msg = Message.newBuilder().setMessageType(MessageType.TP_PROCESS_REQUEST).setContent(req.toByteString())
+        .build();
+
+    TransactionHeader headerNoFamily = TransactionHeader.newBuilder().setFamilyName("someother").setFamilyVersion(testVersion)
+        .build();
+    TpProcessRequest reqNoFamily = TpProcessRequest.newBuilder().setHeader(headerNoFamily).build();
+    Message msgNoFamily = Message.newBuilder().setMessageType(MessageType.TP_PROCESS_REQUEST).setContent(reqNoFamily.toByteString())
+        .build();
+    
+    Message pingMessage = Message.newBuilder().setMessageType(MessageType.PING_REQUEST).build();
+    
+    when(stream.receive()).thenReturn(msg,msgNoFamily,null,pingMessage);
+    Future f=new FutureByteString(RandomString.make(70));
+    try {
+      f.setResult(ByteString.EMPTY);
+    } catch (ValidatorConnectionError exc2) {
+    }
+    when(stream.send(any(), any())).thenReturn(f);
+
+    TransactionHandler handler = mock(TransactionHandler.class);
+    when(handler.transactionFamilyName()).thenReturn(testFamily);
+    when(handler.getVersion()).thenReturn(testVersion);
+    
+    tp.addHandler(handler);
+    ExecutorService svc = Executors.newFixedThreadPool(1);
+    svc.submit(tp);
+
+    // Wait a moment for things to execute before shutting down
+    synchronized (svc) {
+      try {
+        svc.wait(1000);
```

Comment:
This should be made deterministic. We've had a lot of problem with sleeps in tests (both making the tests take far too long overall as well as slow environments causing tests to fail.

Reply:
Agreed, although this isn't as bad as the examples you are referring to.

Reply:
Probably easiest thing to do will require moving the loop body out into a private method which takes care of one turn of the wheel.

Reply:
Fixed in latest commit.